### PR TITLE
Add support for X-RateLimit-* headers in ratelimit filter

### PR DESCRIPTION
**Pull request by Pchelolo**
Signed-off-by: Petr Pchelko <ppchelko@wikimedia.org>

### CODE REVIEW

File: api/envoy/extensions/filters/http/ratelimit/v3/rate_limit.proto [hunk 1]
```diff
@@ -64,4 +64,9 @@ message RateLimit {
   // success.
   config.ratelimit.v3.RateLimitServiceConfig rate_limit_service = 7
       [(validate.rules).message = {required: true}];
+
+  // If set to true enables emitting X-RateLimit-* headers by the filter.
+  // The headers are specified in the `draft-polli-ratelimit-headers-00 RFC
+  // <https://tools.ietf.org/id/draft-polli-ratelimit-headers-00.html>`_
```
Comment (clarakosi) [inline api/envoy/extensions/filters/http/ratelimit/v3/rate_limit.proto]: Should probably include that it defaults to false(?) and get rid of the `_` at the end

File: source/extensions/filters/http/ratelimit/ratelimit.cc [hunk 2]
```diff
@@ -161,6 +164,18 @@ void Filter::complete(Filters::Common::RateLimit::LimitStatus status,
...
   }
 
+  if (config_->enableResponseHeaders()) {
+    Http::ResponseHeaderMapPtr rate_limit_headers =
+        RateLimitHeaders::create(std::move(descriptor_statuses));
+    if (response_headers_to_add_ == nullptr) {
+      response_headers_to_add_ = Http::ResponseHeaderMapImpl::create();
+    }
+    Http::HeaderUtility::addHeaders(*response_headers_to_add_, *rate_limit_headers);
+    rate_limit_headers = nullptr;
```
Comment (clarakosi) [inline source/extensions/filters/http/ratelimit/ratelimit.cc]: interesting. Why are we resetting the rate_limit_headers here and the descriptor_statuses below?

File: source/extensions/filters/http/ratelimit/ratelimit_headers.cc [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  case envoy::service::ratelimit::v3::RateLimitResponse::RateLimit::Unit::UNKNOWN:
+  default:
+    return 0;
+  }
+}
+
+} // namespace RateLimitFilter
+} // namespace HttpFilters
+} // namespace Extensions
+} // namespace Envoy
```
Comment (clarakosi) [inline source/extensions/filters/http/ratelimit/ratelimit_headers.cc]: Missing newline at the end?

File: source/extensions/filters/http/ratelimit/ratelimit_headers.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+private:
+  static uint32_t
+  convertRateLimitUnit(envoy::service::ratelimit::v3::RateLimitResponse_RateLimit_Unit unit);
+};
+
+} // namespace RateLimitFilter
+} // namespace HttpFilters
+} // namespace Extensions
+} // namespace Envoy
```
Comment (clarakosi) [inline source/extensions/filters/http/ratelimit/ratelimit_headers.h]: Missing newline at the end?

File: test/extensions/filters/common/ratelimit/utils.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        statusMsg.mutable_current_limit();
+    limitMsg->set_requests_per_unit(requests_per_unit);
+    limitMsg->set_unit(unit);
+    limitMsg->set_name(name);
+  }
+  return statusMsg;
+}
+
+} // namespace RateLimit
+} // namespace Envoy
```
Comment (clarakosi) [inline test/extensions/filters/common/ratelimit/utils.h]: Missing newline at the end?

File: source/common/http/headers.h [hunk 6]
```diff
@@ -313,6 +316,11 @@ class HeaderValues {
     const std::string Nosniff{"nosniff"};
   } XContentTypeOptionValues;
 
+  struct {
+    const std::string Window{"window="};
+    const std::string Name{"name="};
```
Comment (clarakosi) [inline source/common/http/headers.h]: `name` lets it be pretty broad which is maybe a good thing (I see you used it for "first" and "second"  in the tests which I don't fully understand) but in most cases here wouldn't it be just the unit?

**Review by clarakosi — COMMENTED** [review]
A couple of questions but overall LGTM

File: api/envoy/extensions/filters/http/ratelimit/v3/rate_limit.proto (hunk 1 above)
Comment (Pchelolo) [reply api/envoy/extensions/filters/http/ratelimit/v3/rate_limit.proto]: Done for 'false'. `_` seem to be a part of the external links syntax.

File: source/common/http/headers.h (hunk 6 above)
Comment (Pchelolo) [reply source/common/http/headers.h]: I guess I should rename the const. These are the keys for the policy definition as defined in https://tools.ietf.org/id/draft-polli-ratelimit-headers-00.html#rfc.section.3.1 - the quota-policy can contain arbitrary keys describing it. And in our protocol we have names for the descriptors, so I thought why not use it?

File: source/extensions/filters/http/ratelimit/ratelimit.cc (hunk 2 above)
Comment (Pchelolo) [reply source/extensions/filters/http/ratelimit/ratelimit.cc]: RateLimitHeaders::create(std::move(descriptor_statuses)); on line 168 moved the ownership of the descriptor_statuses pointer to RateLimitHeaders, which is consuming the pointer. if this feature is disabled, we still need to consume the pointer, thus L176. As for rate_limit_headers - we are the owners of it, we have to consume it after using

File: source/extensions/filters/http/ratelimit/ratelimit_headers.cc (hunk 3 above)
Comment (Pchelolo) [reply source/extensions/filters/http/ratelimit/ratelimit_headers.cc]: Done

File: source/extensions/filters/http/ratelimit/ratelimit_headers.h (hunk 4 above)
Comment (Pchelolo) [reply source/extensions/filters/http/ratelimit/ratelimit_headers.h]: done

File: test/extensions/filters/common/ratelimit/utils.h (hunk 5 above)
Comment (Pchelolo) [reply test/extensions/filters/common/ratelimit/utils.h]: done

**Pull request closed without merging**